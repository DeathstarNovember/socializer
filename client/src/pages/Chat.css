@media (min-device-width: 768px) {
  .chat-row {
    height: calc(100vh - 80px);
  }
}

@media (max-device-width: 768px) {
  .row.chat-row {
    position: absolute;
    top: 80px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    margin-left: 0px;
    margin-right: 0px;
  }
}

.chat-col {
  height: 100%;
}

@media (max-device-width: 768px) {
  .chat-row .chat-col {
    padding-left: 0px;
    padding-right: 0px;
  }
}
