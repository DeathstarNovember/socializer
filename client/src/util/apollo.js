import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
