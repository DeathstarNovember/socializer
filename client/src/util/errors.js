import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
