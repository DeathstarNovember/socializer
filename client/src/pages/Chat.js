import React, { useContext } from "react";
import { Col, Container, Row } from "react-bootstrap";
import { Helmet } from "react-helmet";
import { Redirect, Route, Switch } from "react-router-dom";
import cx from "classnames";
import { ChatBar, Conversation } from "components";
import { AuthContext } from "util/context";
import "./Chat.css";

const ChatEmptyMessage = () => (
  <p>Select a conversation from the sidebar to begin chatting</p>
);

const Chat = ({
  match: {
    params: { id },
  },
}) => {
  const { token } = useContext(AuthContext);

  if (!token) {
    return <Redirect to="/" />;
  }

  return (
    <>
      <Helmet>
        <title>Socializer | Chat</title>
        <meta property="og:title" content="Socializer | Chat" />
      </Helmet>

      <Container>
        <Row className="chat-row">
          <Col
            xs={id ? 0 : 12}
            md={4}
            className={cx("chat-col", { "d-none d-md-block": id })}
          >
            <ChatBar />
          </Col>
          <Col
            xs={id ? 12 : 0}
            md={8}
            className={cx("chat-col", { "d-none d-md-block": !id })}
          >
            <Switch>
              <Route path="/chat/:id" component={Conversation} />
              <Route component={ChatEmptyMessage} />
            </Switch>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default Chat;
