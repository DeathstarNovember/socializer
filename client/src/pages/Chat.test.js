import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "util/context";
import Chat from "./Chat";

jest.mock("react-router-dom", () => {
  const { MemoryRouter, Route, Switch } = jest.requireActual(
    "react-router-dom",
  );

  return {
    Redirect: () => <div className="redirect" />,
    MemoryRouter,
    Route,
    Switch,
  };
});

jest.mock("components", () => ({
  ChatBar: () => <div className="chat-bar" />,
  Conversation: () => <div className="conversation" />,
}));

describe("Chat", () => {
  it("redirects when unauthenticated", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/chat/123"]}>
        <AuthContext.Provider value={{}}>
          <Chat match={{ params: { id: 123 } }} />
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when chat is selected", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/chat/123"]}>
        <AuthContext.Provider value={{ token: "abc" }}>
          <Chat match={{ params: { id: 123 } }} />
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when no chat is selected", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ token: "abc" }}>
          <Chat match={{ params: {} }} />
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });
});
