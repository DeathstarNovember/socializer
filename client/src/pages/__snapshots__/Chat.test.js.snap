// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat redirects when unauthenticated 1`] = `
<div>
  <div
    class="redirect"
  />
</div>
`;

exports[`Chat renders correctly when chat is selected 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="chat-row row"
    >
      <div
        class="chat-col d-none d-md-block col-md-4 col-0"
      >
        <div
          class="chat-bar"
        />
      </div>
      <div
        class="chat-col col-md-8 col-12"
      >
        <div
          class="conversation"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Chat renders correctly when no chat is selected 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="chat-row row"
    >
      <div
        class="chat-col col-md-4 col-12"
      >
        <div
          class="chat-bar"
        />
      </div>
      <div
        class="chat-col d-none d-md-block col-md-8 col-0"
      >
        <p>
          Select a conversation from the sidebar to begin chatting
        </p>
      </div>
    </div>
  </div>
</div>
`;
